### PR TITLE
Upgrade to Apache Hdfs 3.2.0

### PR DESCRIPTION
**Pull request by ishanchandrajoshi**
This branch contains the updated code to run the framework with Apache Hdfs 3.2.0. It also contains a temporary fix which modifies the keytab file so as kerby reads it in proper way.

### CODE REVIEW

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/PodInfoBuilder.java [hunk 1]
```diff
@@ -584,7 +584,7 @@ private static String getConfigTemplateDownloadPath(ConfigFileSpec config) {
             podSpec.getSeccompUnconfined(),
             podSpec.getSeccompProfileName());
 
-      if (podSpec.getSeccompUnconfined()) {
+      if (podSpec.getSeccompUnconfined() != null && podSpec.getSeccompUnconfined()) {
```
Comment (kaiwalyajoshi) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/PodInfoBuilder.java]: Lets remove this change from this PR, we'll issue a seperate PR to track breakages due to upgrades relating to Seccomp. (https://github.com/mesosphere/dcos-commons/pull/3070)

File: frameworks/hdfs/tests/test_ssl_kerberos_auth.py [hunk 2]
```diff
@@ -133,7 +133,6 @@ def test_verify_https_ports(hdfs_client, node_type, port):
     ok, stdout, stderr = config.run_client_command("curl -v --cacert {} https://{}".format(ca_bundle, host))
     assert ok
 
-    assert "SSL connection using TLS1.2 / ECDHE_RSA_AES_128_GCM_SHA256" in stderr
```
[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/tests/test_ssl_kerberos_auth.py]: It outputs this, but has changed for JN, NN and DN.
Journal Node: ECDHE_RSA_AES_128_GCM_SHA256
Name Node: ECDHE_RSA_AES_256_CBC_SHA384
Data Node: ECDHE_RSA_AES_256_CBC_SHA384

We are asserting two more statements:
`assert "server certificate verification OK" in stderr`
`assert "common name: {}.{} (matched)".format(task_id, config.SERVICE_NAME) in stderr`
I think these should suffice.

**Comment by samvantran** [comment]
Hey @ishanchandrajoshi it's great that we're finally at a working place. Can you add an upgrade test for HDFS so that we can make sure we automate future upgrade checks? Similar to: https://github.com/mesosphere/dcos-kafka-service/blob/master/frameworks/kafka/tests/test_upgrade.py#L16

**Comment by ishanchandrajoshi** [comment]
Thanks @samvantran, yes indeed after a long struggle we got through all the tests. I have added a test which will check whether Hdfs upgrade works or not. Please check it.

File: frameworks/hdfs/tests/test_upgrade.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+foldered_name = config.FOLDERED_SERVICE_NAME
+
+
+@pytest.mark.sanity
+def test_hdfs_upgrade():
+    try:
+        sdk_install.uninstall(config.PACKAGE_NAME, foldered_name)
+
+        if sdk_utils.dcos_version_less_than("1.9"):
```
Comment (samvantran) [inline frameworks/hdfs/tests/test_upgrade.py]: you can remove this if block. We're only testing on 1.11 and newer these days.

File: frameworks/hdfs/universe/resource.json [hunk 4]
```diff
@@ -3,7 +3,8 @@
     "uris": {
       "jre-tar-gz": "{{jre-url}}",
       "libmesos-bundle-tar-gz": "{{libmesos-bundle-url}}",
-      "hdfs-tar-gz": "https://downloads.mesosphere.com/hdfs/assets/hadoop-2.6.0-cdh5.11.0.tar.gz",
+      "keytab-zip": "https://s3.amazonaws.com/hdfs-binary/krbtest-2.0.tar.gz",
+      "hdfs-tar-gz": "https://s3.amazonaws.com/hdfs-binary/hadoop-3.2.0.tar.gz",
```
Comment (samvantran) [inline frameworks/hdfs/universe/resource.json]: Since we know these tarballs are good, let's move them to `downloads.mesosphere` and update the links.

File: frameworks/hdfs/src/main/dist/modify-keytab.sh [hunk 5]
```diff
@@ -0,0 +1,3 @@
+export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+$JAVA_HOME/bin/java -jar krbtest.jar secrets-hdfs.keytab 
+echo "Keytab file modified"
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/modify-keytab.sh]: Can you change this message to something that describes the fix and reference the HDFS issue? eg
```
echo "Keytab file fixed due to bug in HDFS version 3.2.0. See HDFS#123 http://apache.hdfs<url_link>
```
That way customers know what's happening and can see why we're modifying their keytabs rather than be cryptic about it. Once HDFS releases a new version including the fix we can later delete this script.

Can you also rename the file to something like `fix-keytab-bug.sh`

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 6]
```diff
@@ -213,12 +219,28 @@ pods:
           cp -v zone-resolver.sh {{TASKCFG_ALL_HDFS_VERSION}}/etc/hadoop/zone-resolver.sh
           chmod u+x {{TASKCFG_ALL_HDFS_VERSION}}/etc/hadoop/zone-resolver.sh
           {{/NAME_NODE_PLACEMENT}}
-          ./{{TASKCFG_ALL_HDFS_VERSION}}/bin/hdfs namenode
+          {{#SECURITY_KERBEROS_ENABLED}}
+          sh modify-keytab.sh
+          {{/SECURITY_KERBEROS_ENABLED}}
+          export LATEST_FSIMAGE_VERSION=-65
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: Where did you get this magic number?

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 7]
```diff
@@ -213,12 +219,28 @@ pods:
...
+          if [[ `cat name-data/current/VERSION | grep "layoutVersion" | cut -d'=' -f2` -gt $LATEST_FSIMAGE_VERSION ]]; then
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: can you save the command to a variable that expresses its intent? Also `$()` is the preferred way of wrapping a command over backticks

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 8]
```diff
@@ -213,12 +219,28 @@ pods:
...
+            ./{{TASKCFG_ALL_HDFS_VERSION}}/bin/hdfs namenode
+          fi
         resource-set: name-resources
         readiness-check:
           cmd: |
             set -e
             export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+            export service_state=`./{{TASKCFG_ALL_HDFS_VERSION}}/bin/hdfs haadmin -getServiceState $TASK_NAME | grep -i 'standby'`
+            export LATEST_FSIMAGE_VERSION=-65
+            if [[ $service_state -eq 'standby' && $FINALIZED -eq false && $POD_INSTANCE_INDEX -eq 1 && `cat name-data/current/VERSION | grep "layoutVersion" | cut -d'=' -f2` -gt $LATEST_FSIMAGE_VERSION ]]; then
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: Same here, make the command a var so we know what the last logic check is

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 9]
```diff
@@ -265,21 +291,26 @@ pods:
         cmd: |
           set -e
           export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
-          cp -v hdfs {{TASKCFG_ALL_HDFS_VERSION}}/bin/
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: why'd you delete this and every where else? No longer exists in hdfs vs cdh?

File: frameworks/hdfs/src/main/dist/modify-keytab.sh (hunk 5 above)
Comment (mpereira) [reply frameworks/hdfs/src/main/dist/modify-keytab.sh]: Please add a shebang to the script and `set -x`.

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 10]
```diff
@@ -60,21 +61,26 @@ pods:
...
           {{#SECURITY_KERBEROS_ENABLED}}
           export HADOOP_OPTS="-Djava.security.krb5.conf=$MESOS_SANDBOX/{{TASKCFG_ALL_HDFS_VERSION}}/etc/hadoop/krb5.conf $HADOOP_OPTS"
           {{#TASKCFG_ALL_SECURITY_KERBEROS_DEBUG}}
           export HADOOP_OPTS="-Dsun.security.krb5.debug=true $HADOOP_OPTS"
           {{/TASKCFG_ALL_SECURITY_KERBEROS_DEBUG}}
           {{/SECURITY_KERBEROS_ENABLED}}
           ./bootstrap
+          {{#SECURITY_KERBEROS_ENABLED}}
+          sh modify-keytab.sh
```
Comment (mpereira) [inline frameworks/hdfs/src/main/dist/svc.yml]: Please add the shebang to the script and run it like `./script_name.sh` instead.

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 7 above)
Comment (mpereira) [reply frameworks/hdfs/src/main/dist/svc.yml]: +1

File: frameworks/hdfs/universe/resource.json (hunk 4 above)
Comment (mpereira) [reply frameworks/hdfs/universe/resource.json]: +1

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 11]
```diff
@@ -18,7 +19,7 @@ pods:
     secrets:
       keytab:
         secret: {{SECURITY_KERBEROS_KEYTAB_SECRET}}
-        file: hdfs.keytab
+        file: secrets-hdfs.keytab
```
Comment (mpereira) [inline frameworks/hdfs/src/main/dist/svc.yml]: Why was this renamed?

File: frameworks/hdfs/src/main/dist/modify-keytab.sh [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+$JAVA_HOME/bin/java -jar krbtest.jar secrets-hdfs.keytab
```
Comment (mpereira) [inline frameworks/hdfs/src/main/dist/modify-keytab.sh]: Can we make the keytab an argument for this script?

File: frameworks/hdfs/src/main/dist/modify-keytab.sh (hunk 5 above)
Comment (mpereira) [reply frameworks/hdfs/src/main/dist/modify-keytab.sh]: Also, could you name the script something more specific with regards to its intent? "modify" is very vague imo.

File: frameworks/hdfs/src/main/dist/fix-keytab-bug.sh [hunk 13]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -x
+export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+$JAVA_HOME/bin/java -jar krbtest.jar $1
+echo "Keytab file fixed due to bug in HDFS version 3.2.0. See HADOOP-16283 https://issues.apache.org/jira/plugins/servlet/mobile#issue/HADOOP-16283"
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/fix-keytab-bug.sh]:
```suggestion
echo "Keytab file fixed due to bug in HDFS version 3.2.0. See HADOOP-16283 https://issues.apache.org/jira/browse/HADOOP-16283"
```
Let's not use mobile links

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 14]
```diff
@@ -213,12 +220,31 @@ pods:
...
+            ./{{TASKCFG_ALL_HDFS_VERSION}}/bin/hdfs namenode
+          fi
         resource-set: name-resources
         readiness-check:
           cmd: |
             set -e
             export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+            export service_state=$(./{{TASKCFG_ALL_HDFS_VERSION}}/bin/hdfs haadmin -getServiceState $TASK_NAME | grep -i 'standby')
+            export LATEST_FSIMAGE_VERSION=-65
+            export CURRENT_FSIMAGE_VERSION=$(cat name-data/current/VERSION | grep "layoutVersion" | cut -d'=' -f2)
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: you don't need to cat a file into grep, you can more simply
```
grep "layoutVersion" name-data/current/VERSION | cut -d....
```

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 11 above)
Comment (samvantran) [reply frameworks/hdfs/src/main/dist/svc.yml]: +1 pls revert

**Comment by samvantran** [comment]
Hey @ishanchandrajoshi looks like your new changes failed the flake8 tests. You can run these python checks locally before pushing a commit. Please see https://github.com/mesosphere/dcos-commons/blob/master/TESTING.md#using-pre-commit. Also there were some questions left unanswered. Please take a 2nd look over those.

File: frameworks/hdfs/tests/test_upgrade.py (hunk 3 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/tests/test_upgrade.py]: Done

File: frameworks/hdfs/universe/resource.json (hunk 4 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/universe/resource.json]: Done

File: frameworks/hdfs/src/main/dist/modify-keytab.sh (hunk 5 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/modify-keytab.sh]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 6 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: This is the version of the fsimage in Apache Hadoop 3.2.0. Here is a fsimage converted to XML format from Apache Hadoop 3.2.0:
`<?xml version="1.0"?>
<fsimage><version><layoutVersion>-65</layoutVersion><onDiskVersion>1</onDiskVersion><oivRevision>Unknown</oivRevision></version>
<NameSection><namespaceId>425942755</namespaceId><genstampV1>1000</genstampV1><genstampV2>1000</genstampV2><genstampV1Limit>0</genstampV1Limit><lastAllocatedBlockId>1073741824</lastAllocatedBlockId><txid>0</txid></NameSection>
<ErasureCodingSection>
<erasureCodingPolicy>
<policyId>5</policyId><policyName>RS-10-4-1024k</policyName><cellSize>1048576</cellSize><policyState>DISABLED</policyState><ecSchema>
<codecName>rs</codecName><dataUnits>10</dataUnits><parityUnits>4</parityUnits></ecSchema>
</erasureCodingPolicy>

<erasureCodingPolicy>
<policyId>2</policyId><policyName>RS-3-2-1024k</policyName><cellSize>1048576</cellSize><policyState>DISABLED</policyState><ecSchema>
<codecName>rs</codecName><dataUnits>3</dataUnits><parityUnits>2</parityUnits></ecSchema>
</erasureCodingPolicy>

<erasureCodingPolicy>
<policyId>1</policyId><policyName>RS-6-3-1024k</policyName><cellSize>1048576</cellSize><policyState>ENABLED</policyState><ecSchema>
<codecName>rs</codecName><dataUnits>6</dataUnits><parityUnits>3</parityUnits></ecSchema>
</erasureCodingPolicy>

<erasureCodingPolicy>
<policyId>3</policyId><policyName>RS-LEGACY-6-3-1024k</policyName><cellSize>1048576</cellSize><policyState>DISABLED</policyState><ecSchema>
<codecName>rs-legacy</codecName><dataUnits>6</dataUnits><parityUnits>3</parityUnits></ecSchema>
</erasureCodingPolicy>

<erasureCodingPolicy>
<policyId>4</policyId><policyName>XOR-2-1-1024k</policyName><cellSize>1048576</cellSize><policyState>DISABLED</policyState><ecSchema>
<codecName>xor</codecName><dataUnits>2</dataUnits><parityUnits>1</parityUnits></ecSchema>
</erasureCodingPolicy>

</ErasureCodingSection>

<INodeSection><lastInodeId>16385</lastInodeId><numInodes>1</numInodes><inode><id>16385</id><type>DIRECTORY</type><name></name><mtime>0</mtime><permission>nobody:supergroup:0755</permission><nsquota>9223372036854775807</nsquota><dsquota>-1</dsquota></inode>
</INodeSection>
<INodeReferenceSection></INodeReferenceSection><SnapshotSection><snapshotCounter>0</snapshotCounter><numSnapshots>0</numSnapshots></SnapshotSection>
<INodeDirectorySection></INodeDirectorySection>
<FileUnderConstructionSection></FileUnderConstructionSection>
<SecretManagerSection><currentId>0</currentId><tokenSequenceNumber>0</tokenSequenceNumber><numDelegationKeys>0</numDelegationKeys><numTokens>0</numTokens></SecretManagerSection><CacheManagerSection><nextDirectiveId>1</nextDirectiveId><numDirectives>0</numDirectives><numPools>0</numPools></CacheManagerSection>
</fsimage>`

Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: `layoutVersion` is the version of `fsimage`.

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 7 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 8 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 9 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Latest version of both Apache and CDH also contains `hdfs` script. But, it has completely changed compared to 2.6.0. In the previous version additional logic was applied:
`bin=`dirname $0``
`bin="$MESOS_SANDBOX/${bin}"` in the hdfs script. That's why it was copied here. In the case of latest version, everything works well without applying this logic.

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 10 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Done

File: frameworks/hdfs/src/main/dist/modify-keytab.sh (hunk 12 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/modify-keytab.sh]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 14 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 11 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: There was a difficulty in writing the output in the same file `hdfs.keytab` as the input, in the Java program which modifies the keytab file. That is why the name of input file was changed to `secrets-hdfs.keytab`. The output file name remains the same `hdfs.keytab`.

File: frameworks/hdfs/src/main/dist/fix-keytab-bug.sh [hunk 15]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+set -x
+export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
+$JAVA_HOME/bin/java -jar krbtest.jar $1
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/fix-keytab-bug.sh]: Can you rename the java jar to something more appropriate? `krbtest` doesn't sound right.

Also everywhere you're calling this script already includes 
```
set -x 
export JAVA_HOME=$(ls -d $MESOS_SANDBOX/jdk*/)
```
can we move these two other commands directly into `svc.yml` and get rid of this script entirely?

File: frameworks/hdfs/universe/resource.json [hunk 16]
```diff
@@ -3,8 +3,10 @@
     "uris": {
       "jre-tar-gz": "{{jre-url}}",
       "libmesos-bundle-tar-gz": "{{libmesos-bundle-url}}",
-      "hdfs-tar-gz": "https://downloads.mesosphere.com/hdfs/assets/hadoop-2.6.0-cdh5.11.0.tar.gz",
+      "keytab-zip": "https://downloads.mesosphere.com/hdfs/assets/krbtest-2.0.tar.gz",
+      "hdfs-tar-gz": "https://downloads.mesosphere.com/hdfs/assets/hadoop-3.2.0.tar.gz",
       "hdfs-bin-tar-gz": "https://downloads.mesosphere.com/hdfs/assets/hdfs-bin.tar.gz",
+      "hdfs-bug-issue-uri": "https://issues.apache.org/jira/browse/HADOOP-16283",
```
[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/universe/resource.json]: `build.sh` failed with a error:
`Found a bad URI: issues.apache.org/jira/browse/HADOOP-16283" in: /build/frameworks/hdfs/src/main/dist/svc.yml Export URIs to resource.json to allow packaging for airgapped clusters.
Airgap check FAILED. This framework will NOT work in an airgap. Fix the detected issues.
`

[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/universe/resource.json]: This resource is referenced in `resource.json` and `marathon.json.mustache`, but is not downloaded in sandbox, since it is not included in `-uris` in `svc`.

File: frameworks/hdfs/universe/marathon.json.mustache [hunk 17]
```diff
@@ -60,6 +60,8 @@
     "HDFS_URI": "{{resource.assets.uris.hdfs-tar-gz}}",
     "HDFS_BIN_URI": "{{resource.assets.uris.hdfs-bin-tar-gz}}",
     "HDFS_JAVA_URI": "{{resource.assets.uris.hdfs-jre-tar-gz}}",
+    "KEYTAB_URI": "{{resource.assets.uris.keytab-zip}}",
+    "TASKCFG_ALL_HDFS_BUG_ISSUE_URI": "{{resource.assets.uris.hdfs-bug-issue-uri}}",
```
[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/universe/marathon.json.mustache]: Same reason as above.

File: frameworks/hdfs/src/main/dist/fix-keytab-bug.sh [hunk 18]
```diff
@@ -0,0 +1,5 @@
...
+echo "Keytab file fixed due to bug in HDFS version 3.2.0. See HADOOP-16283 {{HDFS_BUG_ISSUE_URI}}"
```
[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/fix-keytab-bug.sh]: Same as above.

File: frameworks/hdfs/universe/resource.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (mpereira) [reply frameworks/hdfs/universe/resource.json]: Seems like we need to fix the airgap linter instead in this case, and still remove the entry added to the resources as Sam said.

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 6 above)
Comment (mpereira) [reply frameworks/hdfs/src/main/dist/svc.yml]: Please add a comment either adding a URL where one can understand what `-65` means or an actual explanation.

File: frameworks/hdfs/universe/resource.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (samvantran) [reply frameworks/hdfs/universe/resource.json]: Ah I found the issue here. At first I thought, "that's strange. airgap_linter explicitly ignores comments: https://github.com/mesosphere/dcos-commons/blob/master/tools/airgap_linter.py#L33" but then I realized we're `echo`ing. In that case, let's change the echo to a comment. I just tried locally and `./build.sh aws` works as expected.

File: frameworks/hdfs/src/main/dist/svc.yml [hunk 19]
```diff
@@ -441,16 +478,19 @@ pods:
...
           {{#SECURITY_KERBEROS_ENABLED}}
           export HADOOP_OPTS="-Djava.security.krb5.conf=$MESOS_SANDBOX/{{TASKCFG_ALL_HDFS_VERSION}}/etc/hadoop/krb5.conf $HADOOP_OPTS"
           {{#TASKCFG_ALL_SECURITY_KERBEROS_DEBUG}}
           export HADOOP_OPTS="-Dsun.security.krb5.debug=true $HADOOP_OPTS"
           {{/TASKCFG_ALL_SECURITY_KERBEROS_DEBUG}}
           {{/SECURITY_KERBEROS_ENABLED}}
           ./bootstrap
+          {{#SECURITY_KERBEROS_ENABLED}}
+          $JAVA_HOME/bin/java -jar krbtest.jar secrets-hdfs.keytab
+          echo "Keytab file fixed due to bug in HDFS version 3.2.0. See HADOOP-16283 {{HDFS_BUG_ISSUE_URI}}"
```
Comment (samvantran) [inline frameworks/hdfs/src/main/dist/svc.yml]: Let's change this to a comment and move it one line up above the actual command. Something like
```
# Fix keytab file due to bug in HDFS version 3.2.0. See https://issues.apache.org/jira/browse/HADOOP-16283
$JAVA_HOME/bin/java -jar krbtest.jar secrets-hdfs.keytab
```

File: frameworks/hdfs/universe/resource.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (ishanchandrajoshi) [reply frameworks/hdfs/universe/resource.json]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 19 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: Done

File: frameworks/hdfs/src/main/dist/svc.yml (hunk 11 above)
Comment (ishanchandrajoshi) [reply frameworks/hdfs/src/main/dist/svc.yml]: The problem is that `hdfs.keytab` file has ownership of `root` user since it is a `secret`. `nobody` user cannot overwrite, move or copy this file. It has permissions `644`.
`rm: cannot remove 'hdfs.keytab': Device or resource busy`
I have added a comment against `secrets-hdfs.keytab` that it should be changed after Hdfs releases its bug fix.

Comment (samvantran) [reply frameworks/hdfs/src/main/dist/svc.yml]: Okay that's a fair explanation. I don't think the comment directly reflects why it's being changed but I can live with it. We'll just have to address when there's a new release.

**Comment by ishanchandrajoshi** [comment]
> Thanks @ishanchandrajoshi! It was a long journey but we can finally get this over the line checkered_flag

Thank you very much to you too @samvantran for your guidance.
